File: django/x402_django/sign.py
"""X402v1 signing — replicated byte-for-byte from the frozen contract
(docs/api/x402-contract.md) and the proven Python reference
(sdks/fastapi/x402_fastapi/sign.py). Identical hashlib/hmac logic
guarantees cross-SDK conformance.

    canonical = "X402v1" LF METHOD LF path LF timestamp LF nonce LF sha256hex(body)
      - LF = "\\n" (0x0A), never CRLF
      - METHOD uppercased; path is the platform API path POSTed to
      - timestamp = unix seconds, decimal integer
      - body hashed as UTF-8 bytes, lowercase hex
    signature = HMAC-SHA256(secret, canonical), lowercase hex (64 chars)
"""

from __future__ import annotations

import hashlib
import hmac
import time
import uuid

#: Canonical-string scheme version. MUST match the platform's X402ApiKey::SCHEME.
SCHEME = "X402v1"


def canonical(method: str, path: str, timestamp: int, nonce: str, body: str) -> str:
    # bool is a subclass of int — reject it explicitly so True/False can't pose
    # as a unix timestamp. Floats/NaN/str are likewise rejected (contract §1.2).
    if isinstance(timestamp, bool) or not isinstance(timestamp, int):
        raise ValueError(
            f"x402: timestamp must be an integer (unix seconds); got {timestamp!r}"
        )
    body_hash = hashlib.sha256(body.encode("utf-8")).hexdigest()
    return "\n".join(
        [SCHEME, method.upper(), path, str(timestamp), nonce, body_hash]
    )


def sign(secret: str, method: str, path: str, timestamp: int, nonce: str, body: str) -> str:
    return hmac.new(
        secret.encode("utf-8"),
        canonical(method, path, timestamp, nonce, body).encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()


def signed_headers(
    key_id: str, secret: str, method: str, path: str, body: str
) -> dict[str, str]:
    timestamp = int(time.time())
    nonce = str(uuid.uuid4())
    return {
        "X-X402-Key": key_id,
        "X-X402-Timestamp": str(timestamp),
        "X-X402-Nonce": nonce,
        "X-X402-Signature": sign(secret, method, path, timestamp, nonce, body),
    }

File: express/src/sign.ts
import { createHash, createHmac, randomUUID } from "node:crypto";

/** Canonical-string scheme version. MUST match the platform's X402ApiKey::SCHEME. */
export const SCHEME = "X402v1";

/**
 * Frozen signing contract (replicated byte-for-byte across all SDKs):
 *   canonical = "X402v1" LF METHOD LF path LF timestamp LF nonce LF sha256hex(body)
 *   - LF = "\n" (0x0A), never CRLF
 *   - METHOD uppercased; path is the request path the server sees
 *     (includes the "/api" prefix; no query, no host)
 *   - timestamp = unix seconds, decimal integer
 *   - body hashed with SHA-256, lowercase hex
 *   signature = HMAC-SHA256(secret, canonical), lowercase hex (64 chars)
 */
export function canonical(
  method: string,
  path: string,
  timestamp: number,
  nonce: string,
  body: string,
): string {
  if (!Number.isInteger(timestamp)) {
    throw new RangeError(
      `x402: timestamp must be an integer (unix seconds); got ${timestamp}`,
    );
  }
  return (
    SCHEME +
    "\n" +
    method.toUpperCase() +
    "\n" +
    path +
    "\n" +
    timestamp +
    "\n" +
    nonce +
    "\n" +
    createHash("sha256").update(body, "utf8").digest("hex") // body hashed as UTF-8 bytes — matches PHP hash('sha256', $body) on a UTF-8 string
  );
}

export function sign(
  secret: string,
  method: string,
  path: string,
  timestamp: number,
  nonce: string,
  body: string,
): string {
  return createHmac("sha256", secret)
    .update(canonical(method, path, timestamp, nonce, body))
    .digest("hex");
}

export interface SignedHeaders {
  "X-X402-Key": string;
  "X-X402-Timestamp": string;
  "X-X402-Nonce": string;
  "X-X402-Signature": string;
}

export function signedHeaders(
  keyId: string,
  secret: string,
  method: string,
  path: string,
  body: string,
): SignedHeaders {
  const timestamp = Math.floor(Date.now() / 1000);
  const nonce = randomUUID();
  return {
    "X-X402-Key": keyId,
    "X-X402-Timestamp": String(timestamp),
    "X-X402-Nonce": nonce,
    "X-X402-Signature": sign(secret, method, path, timestamp, nonce, body),
  };
}
